iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java

```diff
@@ -161,6 +211,91 @@ private static <L extends LoadingLimits, A extends LoadingLimitsAdder<L, A>> voi
         adder.add();
     }
 
+    static void report(String message, String key, TypedValue typedValue, Reporter reporter) {
+        reporter.report(Report.builder()
+                .withKey(key)
+                .withDefaultMessage(message)
+                .withSeverity(typedValue)
+                .build());
+    }
+
+    static void removeVoltageLevelAndSubstation(VoltageLevel voltageLevel, Reporter reporter) {
+        Optional<Substation> substation = voltageLevel.getSubstation();
+        if (voltageLevel.getConnectableStream().noneMatch(c -> c.getType() != IdentifiableType.BUSBAR_SECTION && c.getType() != IdentifiableType.BUS)) {
+            String vlId = voltageLevel.getId();
+            voltageLevel.remove();
+            report(String.format("Voltage level %s removed", vlId), "voltageLevelRemoved", TypedValue.INFO_SEVERITY, reporter);
+        }
+        substation.ifPresent(s -> {
+            if (Iterables.isEmpty(s.getVoltageLevels())) {
+                s.remove();
+                report(String.format("Substation %s removed", s.getId()), "substationRemoved", TypedValue.INFO_SEVERITY, reporter);
+            }
+        });
+    }
+
+    static LoadingLimitsBag calcMinLoadingLimitsBag(List<LoadingLimits> loadingLimits) {
+        LoadingLimitsBag limit = new LoadingLimitsBag();
+        List<TemporaryLimitsBag> temporaryLimitsBags = new ArrayList<>();
+        Map<Integer, TemporaryLimitsBag> temporaryLimitsByAcceptableDuration = new HashMap<>();
+        double minPermanentLimitValue = Double.MAX_VALUE;
+
+        for (LoadingLimits loadingLimit : loadingLimits) {
+            // find the min permanent limit value
+            if (loadingLimit.getPermanentLimit() < minPermanentLimitValue) {
+                minPermanentLimitValue = loadingLimit.getPermanentLimit();
+            }
+
+            // find the min temporary limit value for each acceptableDuration
+            Collection<LoadingLimits.TemporaryLimit> temporaryLimits = loadingLimit.getTemporaryLimits();
+            for (LoadingLimits.TemporaryLimit temporaryLimit : temporaryLimits) {
+                Integer acceptableDuration = temporaryLimit.getAcceptableDuration();
+                if (!temporaryLimitsByAcceptableDuration.containsKey(acceptableDuration)) {
+                    temporaryLimitsByAcceptableDuration.put(acceptableDuration, new TemporaryLimitsBag(temporaryLimit));
+                }
+                if (temporaryLimit.getValue() < temporaryLimitsByAcceptableDuration.get(acceptableDuration).getValue()) {
+                    temporaryLimitsByAcceptableDuration.put(acceptableDuration, new TemporaryLimitsBag(temporaryLimit));
+                }
+            }
+            temporaryLimitsBags.addAll(temporaryLimitsByAcceptableDuration.values());
+        }
+
+        limit.setPermanentLimit(minPermanentLimitValue);
```

Comment:
Perhaps use a `LoadingLimits` to remove empty constructor et setters ?

Reply:
empty constructors and some setters have been removed